Jetpack Anti-Spam: update Anti-Spam label in sidebar menu item

rcanepa opened the pull request:
## Changes proposed in this Pull Request:

* Add Jetpack_Site_Products class to get the list of products that a site has.
* Change the Jetpack > Akismet menu label if the site has Jetpack Anti-Spam.
* Change the Akismet logo for the Jetpack logo if the site has Jetpack Anti-Spam.

#### Jetpack product discussion

* Ticket: 1181531115069428-as-1183000549671263

#### Does this pull request change what data or activity we track or use?

* No. It doesn't change what we currently track or use.

#### Testing instructions:

**Prerequisites**: a site with Akismet activated (no Jetpack Anti-Spam), and a site with Jetpack Anti-Spam.

* Run this PR locally or use Jetpack Beta Tester plugin.

#### Akismet test (nothing should be different to prod in this case)
* Visit wp-admin of the site with Akismet plugin activated.
* Click on the Jetpack menu item (you'll land on Jetpack Dashboard).
* Verify that in the Jetpack menu, there is an item labeled as `Akismet Anti-Spam`.
* Click on that menu item to visit Akismet's settings page.
* Verify that you see the Akismet logo on the top of the page.

<img width="1020" alt="wpAdminAkismet" src="https://user-images.githubusercontent.com/3418513/88594461-98967f80-d037-11ea-9a5a-34dd9c64e4bb.png">

#### Jetpack Anti-Spam test
* Visit wp-admin of the site with Jetpack Anti-Spam.
* Verify that in the Jetpack menu, there is an item labeled as `Anti-Spam`.
* Click on that menu item to visit Akismet's settings page.
* Verify that you see the Jetpack logo on the top of the page.

![image](https://user-images.githubusercontent.com/3418513/88594436-916f7180-d037-11ea-82e4-300df5581807.png)

#### Proposed changelog entry for your changes:

* Rebrand Anti-Spam product depending on whether the site has Akismet or Jetpack Anti-Spam.

ChaosExAnima commented on class.jetpack-admin.php:
```diff
@@ -51,6 +52,34 @@ private function __construct() {
+			$logo_base64     = base64_encode( $logo->get_jp_emblem_larger() );
+			$logo_base64_url = "data:image/svg+xml;base64,{$logo_base64}";
+			$style           = ".akismet-masthead__logo-container { background: url({$logo_base64_url}) no-repeat .25rem; height: 1.8125rem; } .akismet-masthead__logo { display: none; }";
+			wp_add_inline_style( 'admin-bar', $style );
```
Getting an error in testing:

> Notice: wp_add_inline_style was called incorrectly. Scripts and styles should not be registered or enqueued until the wp_enqueue_scripts, admin_enqueue_scripts, or login_enqueue_scripts hooks. 

Can this be added that to the `admin_enqueue_scripts` hook?

rcanepa replied:
Thanks for catching this error. I fixed it using `admin_enqueue_scripts` as you suggested.

ChaosExAnima commented on class.jetpack-admin.php:
```diff
@@ -51,6 +52,39 @@ private function __construct() {
+	 * Generate styles to replace Akismet logo for the Jetpack logo. It's a workaround until we create a proper settings page for
+	 * Jetpack Anti-Spam. Without this, we would have to change the logo from Akismet codebase and we want to avoid that.
+	 */
+	private static function akismet_logo_replacement_styles() {
```
This should not be static, as that breaks the callback on line 74. Static callables take a different format, see [the docs here](https://www.php.net/manual/en/language.types.callable.php).
```suggestion
	private function akismet_logo_replacement_styles() {
```

rcanepa replied:
TIL! I have updated the PR.

rcanepa replied:
Nevermind. `private` without `static` didn't work well. I'm getting this warning:
```
 PHP Warning:  call_user_func_array() expects parameter 1 to be a valid callback, cannot access private method Jetpack_Admin::akismet_logo_replacement_styles() in /var/www/html/wp-includes/class-wp-hook.php on line 287
```

Wouldn't making the function just `static` work with something like:
```
add_action( 'admin_enqueue_scripts', array( $this, 'akismet_logo_replacement_styles' ) );
```

In the link you gave me, I see an example that looks like: 
```
// An example callback method
class MyClass {
    static function myCallbackMethod() {
        echo 'Hello World!';
    }
}
// Type 3: Object method call
$obj = new MyClass();
call_user_func(array($obj, 'myCallbackMethod'));
```

And from what I understand, `$this` should be an instance of the class, so it should work as the example (I believe).

ChaosExAnima commented on class.jetpack-admin.php:
```diff
@@ -51,6 +52,39 @@ private function __construct() {
+	 * Generate styles to replace Akismet logo for the Jetpack logo. It's a workaround until we create a proper settings page for
+	 * Jetpack Anti-Spam. Without this, we would have to change the logo from Akismet codebase and we want to avoid that.
+	 */
+	private function akismet_logo_replacement_styles() {
```
WP hooks need to be public. Sorry, keep on forgetting things! :(
```suggestion
	public function akismet_logo_replacement_styles() {
```

rcanepa commented on class.jetpack-admin.php:
```diff
@@ -51,6 +52,34 @@ private function __construct() {
+
+		// If the site has Jetpack Anti-Spam, change the Akismet menu label accordingly.
+		$site_products = Jetpack_Site_Products::get();
+		if ( false !== array_search( 'jetpack_anti_spam', array_column( $site_products, 'product_slug' ), true ) ) {
```
Great idea. I like it.

rcanepa commented on class.jetpack-site-products.php:
```diff
@@ -0,0 +1,106 @@
+	 * @return array Active Jetpack products
+	 */
+	public static function get() {
+		// this can be expensive to compute so we cache for the duration of a request.
+		if ( is_array( self::$site_products_cache ) && ! empty( self::$site_products_cache ) ) {
+			return self::$site_products_cache;
+		}
+
+		$site_products             = get_option( self::SITE_PRODUCTS_OPTION, array() );
```
You're right. I'm going to delete this internal cache.

rcanepa commented on class.jetpack-admin.php:
```diff
@@ -51,6 +52,39 @@ private function __construct() {
+		if ( false !== array_search( 'jetpack_anti_spam', array_column( $site_products, 'product_slug' ), true ) ) {
+			// Prevent Akismet from adding a menu item.
+			add_action(
+				'init',
+				function () {
+					remove_action( 'admin_menu', array( 'Akismet_Admin', 'admin_menu' ), 5 );
+				},
+				99
```
This piece of code came from a suggestion I received in `pZbK5-5Rp-p2`.

I'm not very familiar with how the hooks system works, but I tested a couple of options and they didn't work (I probably did something wrong). This is what I tried.

Option 1:
```
add_action(
	'jetpack_admin_menu',
	function () {
		remove_action( 'admin_menu', array( 'Akismet_Admin', 'admin_menu' ), 5 );
	},
	99
);
```

Option 2:
```
add_action(
	'admin_menu',
	function () {
		remove_action( 'admin_menu', array( 'akismet_admin', 'admin_menu' ), 15 );
	},
	99
);
```

In both cases, both menu items were visible (Akismet and Jetpack Anti-Spam).

cleacos replied:
Akismet adds another hook after the `admin_menu` call, so we could use this code:

```
add_action(
	'jetpack_admin_menu',
	function () {
		remove_action( 'jetpack_admin_menu', array( 'Akismet_Admin', 'load_menu' ), 5 );
	},
	5
);
```
We have to remove the hook that calls `load_menu` and set the priority higher (5) than the default (10). So with that, we ensure that the call to the hook happens only in the admin pages instead of the whole of the WordPress site.

cfinke replied:
Instead of `priority`=99 in option 2, I think you'd want `priority`=4.  And I agree that doing it that way is better than hooking into `init`.

cfinke replied:
And the priority needs to match. So it should be: 

```
add_action(
	'admin_menu',
	function () {
		remove_action( 'admin_menu', array( 'akismet_admin', 'admin_menu' ), 5 );
	},
	4
);
```

(untested)

rcanepa replied:
This ^ one worked well so I updated the PR to rely on `admin_menu` instead of `init`. Thank you both!

rcanepa commented:
@cleacos, I replaced `array_search` with `in_array` and removed the unnecessary local cache.

cleacos commented on class.jetpack-admin.php:
```diff
@@ -51,6 +52,39 @@ private function __construct() {
+
+		// If the site has Jetpack Anti-Spam, change the Akismet menu label accordingly.
+		$site_products = Jetpack_Site_Products::get();
+		if ( in_array( 'jetpack_anti_spam', array_column( $site_products, 'product_slug' ), true ) ) {
```
This only detects Jetpack Anti-Spam yearly, we have to detect both periods, monthly and yearly.

```suggestion
		$anti_spam_products = [ 'jetpack_anti_spam_monthly', 'jetpack_anti_spam' ];
		
		if ( ! empty( array_intersect( $anti_spam_products, array_column( $site_products, 'product_slug' ) ) ) ) {
```

rcanepa replied:
Great catch Clemente! I did what you suggested but I had to change the array syntax from `[...]` to `array(...)` because of the linter.

ChaosExAnima commented:
So it looks good, but I do note that the initial purchase shows Akismet in the sidebar for the initial page load, and only after refresh it swaps to Anti-Spam. Perhaps we can pass a query var from Calypso that indicates a refresh is needed?

ChaosExAnima commented:
Hey @jeherve! Can you take a look now that 8.8 is out? This is the last ticket for Anti-Spam v2 and I'd love to close this out! 😄

jeherve commented:
@rcanepa Do you think you'll have the chance to take another pass at this PR before the code freeze?

rcanepa commented:
> @rcanepa Do you think you'll have the chance to take another pass at this PR before the code freeze?

Absolutely, @jeherve. It's on my plans to address this between today and tomorrow.

rcanepa commented:
> Part of me wonders if we should be consolidating things with the Jetpack_Plan class since a lot of the processes are similar. This would avoid duplications, and the number of requests we make to WordPress.com to get information.

@jeherve, I'm tempted to merge both classes into one called `Jetpack_Site_Subscriptions`. It would provide consumers with the site's plans and products and support exactly the same behavior as before. What do you think?

jeherve commented:
I like the idea, although I may not call it Subscriptions, since we already have a feature by that name.

That said, I don't really have a better name, short of "offerings" or "features".

rcanepa commented:
> I like the idea, although I may not call it Subscriptions, since we already have a feature by that name.
> 
> That said, I don't really have a better name, short of "offerings" or "features".

Good point. Giving it a second thought, `Jetpack_Plan` doesn't sound bad to me, meaning we could leave the name as it's today and add a new public method called `get_products` to return the site's products. Easy to do, won't break existing code, and we can delete the `Jetpack_Site_Products` class entirely. I'll ping you back when it's ready.

rcanepa commented:
@jeherve, I updated and rebased the PR. The summary of the changes is:
* Merge the `Jetpack_Site_Products` class into the `Jetpack_Plan` class.
* Make sure to check if the `Akismet_Admin` class exists before doing anything with it.
* Remove everything that mentioned the `Jetpack_Site_Products` class or its file.
